File: 백준/2306_유전자/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <memory.h>

using namespace std;

class Solution {
    string s;
    size_t N;
    int32_t dp[501][501];

    void init() {
        cin >> s;
        N = s.length();
        memset(dp, 0x00, sizeof(dp));
    }

    void simulate() {
        // X가 KOI 유전자라면, 
        // aXt , gXc 가 KOI 유전자
        // KOI 유전자 둘을 연결해도 KOI 유전자
        // KOI 유전자 최대 길이 구하기
        // dp[i][j] = > i~-j 문자열에서 최대 길이
        for(int stride = 1 ; stride < N ; stride++) {
            for(int start = 0 ; start + stride < N ; start++) {
                int end = start + stride;

                if( (s[start] == 'a' && s[end] == 't' ) || (s[start] == 'g' && s[end] == 'c' )) {
                    dp[start][end] = dp[start+1][end-1] + 2;
                }

                for(int mid = start ; mid < end ; mid++) {
                    dp[start][end] = max( dp[start][end], dp[start][mid] + dp[mid+1][end] );
                }
            }
        }
    }

    void print() {
        cout << dp[0][s.length()-1] << endl;
    }

public:
    void run() {
        init();
        simulate();
        print();
    }
};

int main(void)
{
    Solution sol;
    sol.run();
    return 0;
}

File: 백준/2306_유전자/Main.java
import java.lang.reflect.Array;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

class Solution {

    private String dna;


    public Solution(BufferedReader reader) throws IOException {
        dna = reader.readLine();
    }

    private long bottomUp() {
        // at, gc both are KOI
        // if X is KOI and Y is KOI then XY also KOI
        // if X is KOI then aXt,gXc also KOI
        //
        int N = dna.length();
        int[][] dp = new int[N+1][N+1];

        for ( int stride = 1 ; stride < N ; stride++ ) {
            for ( int start = 0 ; start + stride < N ; start++ ) {
                int end = start + stride;
                char fch = dna.charAt(start);
                char lch = dna.charAt(end);

                if( ( fch == 'a' && lch == 't' ) || ( fch == 'g' && lch == 'c' ) ) {
                    dp[start][end] = dp[start + 1][end - 1] + 2;
                }

                for(int mid = start ; mid < end ; mid++) {
                    dp[start][end] = Math.max(dp[start][end], dp[start][mid] + dp[mid+1][end]);
                }
                
            }
        }

        return dp[0][N-1];
    }


    public void run () {
        StringBuilder sb = new StringBuilder();
        sb.append(bottomUp());
        System.out.println(sb.toString());
    }
}

